File: typescript/src/equipment/modification-journal.ts
/**
 * Resolve a personal-weapon modification from its journal spelling to its exact recipe.
 *
 * The journal omits the technology suffix from Greater Range, Headshot Damage and Higher
 * Accuracy. Their Kinetic, Laser and Plasma recipes cost different materials, so the
 * equipped weapon is needed to settle which recipe the journal means. This mirrors the
 * collision resolver used by ship engineering.
 *
 * @packageDocumentation
 */

import { requireString, requireStringIfPresent } from '../internal/argument-guards.js';
import { normalizeKey } from '../internal/registry-index.js';
import { PERSONAL_MODIFICATION_JOURNAL_NAMES } from './internal/modification-journal-names.js';
import { getPersonalWeaponBySymbol } from './weapons.js';

/**
 * Resolve the modification recipe a journal symbol identifies on a handheld weapon.
 *
 * @param weaponSymbol - Frontier handheld-weapon symbol. A nullish or unknown weapon
 * cannot resolve a collision and leaves `journalSymbol` unchanged.
 * @param journalSymbol - Symbol from the journal modification list. Matching ignores
 * case and surrounding whitespace.
 * @returns The recipe key used by `PERSONAL_MODIFICATIONS` and
 * `PERSONAL_MODIFICATION_COSTS`, or `journalSymbol` unchanged when no collision resolves.
 * @throws {TypeError} If `journalSymbol` is not a string, or if a present weapon symbol
 * is not a string.
 * @example
 * ```ts
 * import { resolvePersonalModificationForWeapon } from '@elite-dangerous-almanac/core/equipment/modification-journal';
 * resolvePersonalModificationForWeapon(
 *   'wpn_m_assaultrifle_kinetic_fauto',
 *   'weapon_range',
 * ); // -> 'weapon_range_kinetic'
 * ```
 */
export function resolvePersonalModificationForWeapon(
    weaponSymbol: string,
    journalSymbol: string,
): string {
    requireStringIfPresent(weaponSymbol, 'resolvePersonalModificationForWeapon: weaponSymbol');
    const wanted = normalizeKey(
        requireString(journalSymbol, 'resolvePersonalModificationForWeapon: journalSymbol'),
        'resolvePersonalModificationForWeapon: journalSymbol',
    );
    const weapon = getPersonalWeaponBySymbol(weaponSymbol);
    if (!weapon) return journalSymbol;

    for (const [recipeSymbol, writtenSymbol] of Object.entries(
        PERSONAL_MODIFICATION_JOURNAL_NAMES,
    )) {
        if (
            writtenSymbol.toLowerCase() === wanted &&
            recipeSymbol.endsWith(`_${weapon.engineeringType}`)
        ) {
            return recipeSymbol;
        }
    }
    return journalSymbol;
}

File: dotnet/src/EliteDangerousAlmanac/Equipment/ModificationJournal.cs
using System;
using System.Collections.Generic;
using EliteDangerousAlmanac.Internal;

namespace EliteDangerousAlmanac.Equipment;

/// <summary>Resolves the modification symbol a journal writes onto its own recipe.</summary>
/// <remarks>
/// The journal leaves the technology off Greater Range, Headshot Damage and Higher
/// Accuracy: it writes <c>weapon_range</c> for the kinetic, laser and plasma recipes alike.
/// Those recipes ask for different materials, so the weapon the symbol was read from is
/// what settles which one the journal means. Its engineering type picks the recipe out of
/// the group. This is the same collision the ship engineering resolver answers.
/// </remarks>
public static class ModificationJournal
{
    private static readonly Lazy<IReadOnlyDictionary<string, string>> JournalNames = new(
        () => SharedData.Load<Dictionary<string, string>>(
            "data/equipment/modification-journal-names.jsonc"));

    /// <summary>Resolves one journal modification symbol against the weapon that carries it.</summary>
    /// <param name="weaponSymbol">The weapon's item symbol.</param>
    /// <param name="journalSymbol">The modification symbol as the journal writes it.</param>
    /// <returns>
    /// The recipe symbol the catalogue is keyed by, or the journal symbol unchanged where no
    /// weapon carries the symbol, or the journal spelling names one recipe already.
    /// </returns>
    /// <exception cref="ArgumentNullException">The journal symbol is absent.</exception>
    public static string ResolveForWeapon(string? weaponSymbol, string journalSymbol)
    {
        if (journalSymbol is null) throw new ArgumentNullException(nameof(journalSymbol));

        string wanted = journalSymbol.Trim();
        PersonalWeapon? weapon = PersonalWeaponCatalogue.FindBySymbol(weaponSymbol);
        if (weapon is null) return journalSymbol;

        string suffix = "_" + weapon.EngineeringType.ToString().ToLowerInvariant();
        foreach (KeyValuePair<string, string> entry in JournalNames.Value)
        {
            if (string.Equals(entry.Value, wanted, StringComparison.OrdinalIgnoreCase)
                && entry.Key.EndsWith(suffix, StringComparison.OrdinalIgnoreCase))
            {
                return entry.Key;
            }
        }

        return journalSymbol;
    }
}
